### ISSUE TITLE
[FR] Extending tag-feature with YAML tags in Commonmark

### ISSUE THREAD

**Issue by Offerel** [pbek/QOwnNotes#906]
#### Expected behaviour                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
Using tags additionally inside the markdown file itself with the help of YAML. For example, if you add at the beginning of a document the tags in the form:
```
---
tags: tag1 tag2 tagn 
...
```
you have added 3 tags, seperated by space in a YAML like structure. This is also a new feature in Epsilon Notes on Android. You can find more infos at http://epsilonexpert.com/e/info/tags.php 
#### Actual behaviour                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
tagging live only in a seperate file/database

**Comment by pbek** [comment]
Tagging is fully scriptable. Search for Tag in the script Repository in
qownnotes and you will find good example to start with

**Comment by Offerel** [comment]
Yes, i know. But the problem is, that i will only see the tags in editor mode, but not in the preview. That's also the case for all other YAML info's in the header. They should be invisible in a preview, but visible in a editor. Sorry, that was not clear in my request. I was to excited about this new feature in Epsilon Notes.

**Comment by pbek** [comment]
You could hide the code with the help of the hook: http://docs.qownnotes.org/en/develop/scripting/README.html#notetomarkdownhtmlhook

**Comment by Offerel** [comment]
Ahh..., i see. I didn't have the time to read it now completely, but i think i get the idea. I will try to implement a YAML parser in my PrimitiveNotes Plugin for Roundcube in the next days and add a script for QOwnNotes in the next release of the plugin.

**Comment by Offerel** [comment]
OK. I have tested something. This is the current state of work:

```
import QtQml 2.0

/**
 * This script hides YAML headers in the notes preview
 */
QtObject {
    /**
     * This function is called when the markdown html of a note is generated
     *
     * It allows you to modify this html
     * This is for example called before by the note preview
     *
     * @param {NoteApi} note - the note object
     * @param {string} html - the html that is about to being rendered
     * @return {string} the modfied html or an empty string if nothing should be modified
     */
    function noteToMarkdownHtmlHook(note, html) {
		var yaml_beginn = html.indexOf("body><hr");
		var yaml_end = html.indexOf("...<");

		if (yaml_end > 0 && yaml_beginn > 0) {
			var yaml = html.substr(yaml_beginn + 15, yaml_end - yaml_beginn - 15);
			var yaml_arr = yaml.split(/\r?\n/);
			yaml_arr.forEach(function(element) {
				if(element.length > 0) {
					var tag = element.substr(0,element.indexOf(":"));
					switch(tag) {
						case "tags":
							var tags = element.substr(element.indexOf(":") +1).split(" ");
							tags.forEach(function(stag) {
								script.tagCurrentNote(stag);
							});
							break;
						case "title":
							console.warn(element.substr(element.indexOf(":") +1));
							break;
						case "toc":
							console.warn(element.substr(element.indexOf(":") +1));
							break;
						case "date":
							console.warn(element.substr(element.indexOf(":") +1));
							break;
						default:
							console.warn(element.substr(element.indexOf(":") +1));
					}
				}
			});
			return html.substr(0, yaml_beginn + 5) + html.substr(yaml_end + 7);
		}
    }
}
```
The above script hides the YAML header and processes some of the tags. Specially, when i found "tags:" i try to add them also as QOwnNotes tags. Technically it works, but when there are many tags, it could take a long time to add tag for tag. I don't know if there is a faster way.  The relevant part is in the line `script.tagCurrentNote(stag);`

**Comment by pbek** [comment]
If you want to take over the whole tagging mechanism you can use: http://docs.qownnotes.org/en/develop/scripting/README.html#notetagginghook

Or you can try if `addTag` from http://docs.qownnotes.org/en/develop/scripting/README.html#note has a better performance for you. Tagging with `tagCurrentNote` causes the user interface to be reloaded (just like adding a tag manually with the button in the user interface).

> why do i get the console.warn twice for every note?

do you get two outputs if you just execute one `console.warn` in an empty script?

**Comment by Offerel** [comment]
Yes, i get the two outputs mostly every time, i click on a new note. Maybe we should move this to a new issue. But first here is what i have done:

* created a new script test.qml
* added the following code in to it:
```
import QtQml 2.0

QtObject {
    function noteToMarkdownHtmlHook(note, html) {
		console.warn("test");
    }
}  
```
* disabled all other scripts
* enabled test.qml alone
* click on a new note
* look at the console output and get:
```
[12:35:24] [warning] test
[12:35:29] [warning] test
```

regarding note tagging, i try to find the right way.

**Comment by Offerel** [comment]
Back to the topic: I tried to understand noteTaggingHook but the example in note-tagging.qml honestly confuses me very much. According to the description, this script finds text passages in a note with an @sign and adds them as a tag to a note.

Do I have to include exactly this file (note-tagging.qml) as script to be able to call noteTaggingHook? Because when I call `noteTaggingHook(note, 'list')`, I get the error message that it does not know noteTaggingHook. Or do I have to create my own noteTaggingHook function in my script and note-tagging. qml is just an example of what this function might look like? I assume that's the last of it. So I would have to check in my previous script if TagXY0815 is already assigned to the note. If not, I add the tag with "add". However, if a tag does not appear in my list of "tags:", I can remove it from the note with "remove".

I've been added note-tagging.qml testwise. If I do this, however, the tags of all notes are already processed when the program is started or at least when the Scripting Engine is restarted. So all I had to do, is to edit note-tagging.qml so, that instead of tags in the form of @tag1 or @tag2, the tags come from an array of "tags:" and hide the YAML header in the preview... im a little bit lost :-)

**Comment by pbek** [comment]
`noteTaggingHook` is for completely taking over note tagging in a script (and don't use the internal tag database of QOwnNotes). You just have to create this one `noteTaggingHook` method in any script and implement what to do when notes are added, renamed, removed or listed by users via the user interface. This is what the big switch statement does. You can implement your own logic, for example in the `add` case that the new tag is put at the right place in the note... The same you do for the other three cases.

**Comment by pbek** [comment]
If you still want to use QOwnNotes tag database then noteTaggingHook is not good for your use-case.

**Comment by pbek** [comment]
> noteToMarkdownHtmlHook

that may just tell us that `noteToMarkdownHtmlHook` is called twice in the process... which could be the case... it's called a lot of times... every time something in the note is changed... you shouldn't tag a note in that hook, especially if it was already tagged.

the first question is if you still want to use QOwnNotes' tag database or want to implement your own tagging in `noteTaggingHook`

**Comment by Offerel** [comment]
ahhh... ok... to decide if i use QOwnNotes tag database or my own isn't that simple. It's not clear to my what are the benefits from using either QOwnNotes tag database or my own thing. But i assume there is a good reason to use QOwnNotes tag database, so, without to know the benefits, i would like to tend to use QOwnNotes tag database.

**Comment by pbek** [comment]
> But i assume there is a good reason to use QOwnNotes tag database, so, without to know the benefits, i would like to tend to use QOwnNotes tag database.

Your script only works with the current note, if that's enough for you that's ok. 
noteTaggingHook is used to check every note in your note folder...

**Comment by pbek** [comment]
If you want you can put your script to <https://github.com/qownnotes/scripts> (as pull request), but be sure to explain what it does. :)

**Comment by pbek** [comment]
And good work for finding a working solution yourself!

**Comment by Maboroshy** [comment]
I was looking into making QON and Epsilon Notes tags compatible too and found this issue.

I'll try using in-text tagging approach rather than db. That way tags will update on each note folder opening rather than individual notes opening. Also I'm going to make QON recreate `Epsilon/tags.dat` file on each tag update to resolve the same "tag update on note open" issue of Epsilon Notes.

**Comment by Maboroshy** [comment]
I've uploaded "Epsilon Notes compatible tags (YAML tags)" script which uses in-text tagging to repository. Recreating `Epsilon/tags.dat` is planned for future version.

**Comment by Maboroshy** [comment]
Hiding YAML section by `noteToMarkdownHtmlHook` was quite an exercise. The same hook but affecting `noteText` heading to converter instead of `html` would make it much easier.

**Comment by Offerel** [comment]
You hiding part looks great, a lot better than mine. I hope, it's OK for you, when I take your approach for my script also?

**Comment by Maboroshy** [comment]
@pbek I'm stuck with recreation of `Epsilon/tags.dat` which is a text file with tag-note dictionary. I want to create it by `script.writeToFile` after `noteTaggingHook`'s "list" command. I can write the file on each note "list" but that's not very efficient. Is there any way to do it only once, after "listing" all notes is complete?

**Comment by pbek** [comment]
Why do you want to write the file after `list`? `list` is just for reading. Why not after add/remove/rename?

**Comment by Maboroshy** [comment]
I need to create initial `tags.dat` with all the tags-notes pairs in it. I'll can't get all notes tag data during tag changes. I've just thought that maybe I even can't get all notes tag data on "list" since the script relaunches with "list" command for each note losing all the vars. 

For now it looks like good old db ripping from Python is the way to go.

**Comment by pbek** [comment]
Can you do this in the `init` function?
Do you want to transfer all current tags from the QOwnNotes tags database to your tags.dat?

**Comment by Maboroshy** [comment]
I've tried defining var in `init` but hook function does not see it (or I messed up something). Property values also not stored between script "sessions". There's persistent storage but that will be same rewriting disk data on each note "list".

At first I need to get the tag data from all the notes YAML. Then I can either gather tags into array while "listing" and flush the array into `tags.dat` after it's complete, either use the db with "listing" results directly. But I'm not sure what hook should I use to know when db has listing results, since "list" can't tell it's completed. Only changing tag is guaranteed. That's not pretty for the very first run but acceptable.

**Comment by pbek** [comment]
I think I still don't understand. Do you want to migrate QOwnNotes tags to your own tags.dat? That needs to be a separate script (or just a custom action) that is done once only. You could also do that in `init` and store a persistent variable that you already did it...

the then you only interact with your tags.dat in the noteTaggingHook like it is meant to be used (all requests concerning tags will be made to the tags.dat, tagging data will only be present there and not inside QOwnNotes)

**Comment by Maboroshy** [comment]
After some tinkering looks like I made property saving work. I'll investigate further.

**Comment by Maboroshy** [comment]
- Lets presume I have a set of notes with tags in YAML section in each note text.
- I already have YAML tags working via `noteTaggingHook`
- Android's Epsilon notes fills `tags.dat` from note YAML only note by note, on each note opening. It doesn't scan all notes like QON does. I'll have to through each note of my set in Epsilon Notes.
- I don't find it exciting so I want QON to make `tags.dat` with the tags of my set of notes instead.
- I don't want to make `tags.dat` the source of tag data since I think that it's less portable. I want to make YAML in note text as the source of tag data and my script already does that.

Looks like I've made saving to property work. Now I'll look for a proper hook to write `tags.dat`.

**Comment by pbek** [comment]
what is stored in the tags.dat and why can't it be stored in the notetagginghook while add/remove/rename?

**Comment by pbek** [comment]
sorry for not being of great help (yet?) :)

**Comment by Maboroshy** [comment]
example of `tags.dat`:
```
/storage/emulated/0/notes/note1.md
tag1
/storage/emulated/0/notes/note2.md
tag1 tag2 tag3
```
For example lets assume I've put `script.writeToFile('tags.dat', arrayWithAllTags)` to each add/remove/rename.
- First I have to fill arrayWithAllTags. 
- I don't know if it's possible to get all the note path to tag data like some kind of property. I never did that and docs doesn't mention something like that directly.
- I can do that while "listing". When it's complete I'll have my array full and ready.
- I ran QON with the set of notes, QON "listed" them, all my tags are in db, but I'm not done yet since `tag.dat` is empty. It waits for me to start some tagging. I **need** to do tagging to complete my sync routine.

So as I wrote earlier that's not pretty for the very first run and can be confusing for the user.

> sorry for not being of great help (yet?) :)

It's more than a year while I'm here and still under impression how great and patiently you treat your users. Even if they are mumbling with a semi-english brain dumps like me. :stuck_out_tongue_closed_eyes:

**Comment by pbek** [comment]
Hehe, thank you for your kind words. But you are contributing a lot!

Can't you just add/rename/remove single tags for one note in tags.dat with the add/rename/remove handlers of the notetagginghook? I'm not sure yet why you want to rebuild the whole tags.dat if you don't have information about all note tag relations.

**Comment by Maboroshy** [comment]
- First lets make it clear that the target is having all tag data sources in sync. That is note text YAML, QON db and Epsilon notes' tags.dat in sync.
- There are 3 sources any of which can naturally start to differ. 
- Since all of them can conflict one have to be the base one, an anchor. I chose YAML as one since I find it most portable. If one of sources is missing, like tags.dat in our case, it should be recreated from the base source.
- If QON and Epsilon Notes both will add tag data one by one it will take a long time for them to "meet". And any 3rd player that alter some data will restart this journey.
- So there must be a single and transparent moment when all these sources are synced.
- Since there's not much Epsilon Notes can do complete sync should either be done by QON or by some other middle-man that has access to all tag data sources, like a Python script.

**Comment by pbek** [comment]
Since you don't even know when the tags will start to loose their "sync" you could maybe do a complete resync from QOwnNotes to tags.dat in a periodic timer function in the QOwnNotes tagging script?

**Comment by Maboroshy** [comment]
How do I implement a background functions in a QML script? I've never done something like that. In Python I'd start a new thread.

**Comment by Maboroshy** [comment]
I have an illusion that the solution of the whole issue is something like implementing "finally" command for `noteTaggingHook` that runs after "list" is complete and QON tag db s filled.

**Comment by pbek** [comment]
"list" can be called on every note every time a the tag tree is rebuild, that can happen quite a lot...
Why don't you bild your tags.dat with a timer (see <https://github.com/pbek/QOwnNotes/blob/b05aa3089fc68dc9e0ca44daa4ddd3d6d8e65bfe/doc/scripting/scripting-label-demo.qml> periodically.
Or maybe only once after like 10 sec if you prefer..

**Comment by Maboroshy** [comment]
Won't this timer hold whole tagging since it's all a single thread? Or it creates a new thread by itself?

Next stop. I fill a tag data array by "list", keep it up to date on tag changes, when the timer triggers script checks if the up to date array differs from array previously flushed into tags.dat and write the file if it does. Theoretically timer can be triggered during "list" and produce incomplete tags.dat. Chances are very low, but still present. Maybe I should write and test this solution.

**Comment by pbek** [comment]
ad thread: I can't tell for sure... there is a lot of "magic" going on with Qt threads.
yeah, best try and see if there are troubles

**Comment by pbek** [comment]
ad thread: I've never tested this, but maybe that's something for you: <http://doc.qt.io/archives/qt-5.10/qml-workerscript.html>

**Comment by Maboroshy** [comment]
I've a have a working prototype with a timer. It works as expected. Working on tag storage logic now. I basically has to make a parallel tag storage using `noteTaggingHook` feeds. Me and JS/QML have a mutual dislike so all this doesn't come fast.

**Comment by pbek** [comment]
Still sounds great! :)
